src/couch/src/couch_key_tree.erl

```diff
@@ -470,6 +471,53 @@ map_leafs_simple(Fun, Pos, [{Key, Value, SubTree} | RestTree]) ->
 
 
 stem(Trees, Limit) ->
+    lists:sort(lists:flatmap(fun(Tree) ->
+        stem_tree(Tree, Limit)
+    end, Trees)).
+
+
+stem_tree({Depth, Child}, Limit) ->
+    case stem_tree(Depth, Child, Limit) of
+        {_, NewChild, NewBranches} ->
+            [{Depth, NewChild} | NewBranches];
+        {_, NewBranches} ->
+            NewBranches
+    end.
+
+
+stem_tree(_Depth, {_Key, _Val, []} = Leaf, Limit) ->
+    {Limit - 1, Leaf, []};
+
+stem_tree(Depth, {Key, Val, Children}, Limit) ->
+    FinalAcc = lists:foldl(fun(Child, {LimitPosAcc, ChildAcc, BranchAcc}) ->
+        case stem_tree(Depth + 1, Child, Limit) of
+            {LimitPos, NewChild, NewBranches} ->
+                NewLimitPosAcc = erlang:max(LimitPos, LimitPosAcc),
+                NewChildAcc = [NewChild | ChildAcc],
+                NewBranchAcc = NewBranches ++ BranchAcc,
+                {NewLimitPosAcc, NewChildAcc, NewBranchAcc};
+            {LimitPos, NewBranches} ->
+                NewLimitPosAcc = erlang:max(LimitPos, LimitPosAcc),
+                NewBranchAcc = NewBranches ++ BranchAcc,
+                {NewLimitPosAcc, ChildAcc, NewBranchAcc}
+        end
+    end, {-1, [], []}, Children),
+    {FinalLimitPos, FinalChildren, FinalBranches} = FinalAcc,
+    case FinalLimitPos of
+        N when N > 0, length(FinalChildren) > 0 ->
+            FinalNode = {Key, Val, lists:reverse(FinalChildren)},
+            {FinalLimitPos - 1, FinalNode, FinalBranches};
+        0 when length(FinalChildren) > 0 ->
+            NewBranches = lists:map(fun(Child) ->
+                {Depth + 1, Child}
+            end, lists:reverse(FinalChildren)),
+            {-1, NewBranches ++ FinalBranches};
```

Comment:
You could combine the `reverse` and the `++` using `reverse/2`. Can't imagine it matters.

Reply:
Good call.